packages/teleterm/src/services/config/configStore.ts

```diff
@@ -8,58 +8,70 @@ const logger = new Logger('ConfigStore');
 export function createConfigStore<
   Schema extends z.AnyZodObject,
   Shape = z.infer<Schema>
->(schema: Schema, defaults: Shape, fileStorage: FileStorage) {
-  const parsed = parse(fileStorage.get());
-  updateConfigFile();
+>(schema: Schema, fileStorage: FileStorage) {
+  const { storedConfig, configWithDefaults, errors } = validateStoredConfig();
 
   function get<K extends keyof Shape>(
     key: K
   ): { value: Shape[K]; metadata: { isStored: boolean } } {
-    const stored = fileStorage.get<Shape>();
     return {
-      value: { ...defaults, ...stored }[key],
-      metadata: { isStored: stored[key] !== undefined },
+      value: configWithDefaults[key],
+      metadata: { isStored: storedConfig[key] !== undefined },
     };
   }
 
   function set<K extends keyof Shape>(key: K, value: Shape[K]): void {
-    fileStorage.putKey(key as string, value);
+    fileStorage.put(key as string, value);
+    configWithDefaults[key] = value;
+    storedConfig[key] = value;
   }
 
-  function getParsingErrors(): ZodIssue[] | undefined {
-    if (parsed.success === false) {
-      return parsed.error.issues;
-    }
+  function readValidationErrors(): ZodIssue[] | undefined {
+    return errors;
   }
 
-  function parse(data: Shape) {
-    return schema
-      .partial() // make all keys optional
-      .safeParse(data);
+  function parse(data: Partial<Shape>) {
+    return schema.safeParse(data);
   }
 
-  function updateConfigFile(): void {
+  //TODO (gzdunek): syntax errors of the JSON file are silently ignored, report
+  // them to the user too
+  function validateStoredConfig(): {
+    storedConfig: Partial<Shape>;
+    configWithDefaults: Shape;
+    errors: ZodIssue[] | undefined;
+  } {
+    const storedConfig = fileStorage.get<Partial<Shape>>();
+    const parsed = parse(storedConfig);
     if (parsed.success === true) {
-      fileStorage.put(parsed.data); // parsed.data does not contain unknown keys
+      return {
+        storedConfig,
+        configWithDefaults: parsed.data as Shape,
```

Comment:
Something about having to use `as` here rubbed me the wrong way since zod claims to be a TypeScript-first library.

Turns out we can change the definition of Schema to this:

```typescript
export function createConfigStore<
  Schema extends z.ZodTypeAny,
```

and then we can drop the `as` here and by `reParsed`. From what I checked, we still get properly inferred types.

See https://github.com/colinhacks/zod/tree/v3.20.2#writing-generic-functions

Reply:
Thanks for the link, I was sure I missed some obvious explanation how to write functions like this :) 

Changed.

Reply:
just a note here, zod rules!